src/bin/pg_autoctl/pgctl.c

```diff
@@ -334,7 +334,51 @@ config_find_pg_ctl(PostgresSetup *pgSetup)
 	}
 	else if (n == 0)
 	{
-		log_warn("Failed to find pg_ctl in PATH");
+		/*
+		 * When using debian/ubuntu packaging then pg_config is installed as
+		 * part as the postgresql-common package in /usr/bin, whereas pg_ctl is
+		 * installed in a major version dependent location such as
+		 * /usr/lib/postgresql/12/bin, and those locations are not included in
+		 * the PATH.
+		 *
+		 * So when we can't find pg_ctl anywhere in the PATH, we look for
+		 * pg_config instead, and then use pg_config --bindir to discover the
+		 * pg_ctl we can use.
+		 */
+
+		char **pg_configs = NULL;
+		int pg_configs_count = search_path("pg_config", &pg_configs);
+
+		log_debug("Failed to find pg_ctl in PATH, now looking for pg_config");
+
+		if (pg_configs_count == 1)
+		{
+			if (!set_pg_ctl_from_config_bindir(pgSetup, pg_configs[0]))
+			{
+				/* errors have already been logged */
+				return false;
+			}
+
+			if (!pg_ctl_version(pgSetup))
+			{
+				log_fatal("Failed to get version info from %s --version",
+						  pgSetup->pg_ctl);
+				return false;
+			}
+
+			log_debug("Found pg_ctl for PostgreSQL %s at %s from pg_config "
+					  "found in PATH at \"%s\"",
+					  pgSetup->pg_version, pgSetup->pg_ctl, pg_configs[0]);
+
+			/* set our return value to 1 in that case, indicating success */
+			n = pg_configs_count;
+		}
+		else
+		{
+			log_warn("Failed to find either pg_ctl or pg_config in PATH");
```

Comment:
Might be good to have another message in case there's more than 1 `pg_config` found. I don't expect this to happen in real setups though, i.e. no `pg_ctl` but multiple `pg_config`

Reply:
Yeah that's fair. This also triggered a more extensive review of the patch, and I went as far as cleaning up the memory allocation scheme of the `search_path` function, which landed early in the development and did not get the memo about our favorite way to allocate memory (struct on the stack).

Also I exposed the facility to be easily tested for both contexts of a monitor and a keeper:

```
$ PG_AUTOCTL_DEBUG=1 pg_autoctl do pgsetup pg_ctl --pgdata /tmp/m
```

The PGDATA directory does not need to exist, but the command requires one to be known still, because it's sharing the same getops command as many other commands in the `pg_autoctl do` family.

Finally, we currently have a different processing of `--pgctl` in the monitor and the keeper. That's because on the monitor we're not picky about the Postgres version to use, whereas on the monitor we really insist on there being absolutely no ambiguity. That said, I have to wonder if we shouldn't just apply the same thoroughness checks in both cases, actually?